GitUpKit/Interface/GIGraphView.m

```diff
@@ -860,112 +860,170 @@ static void _DrawLine(GILine* line, CGContextRef context, CGFloat offset, CGFloa
+    [boldRanges addObject:[NSValue valueWithRange:branchNameRange]];
+    [darkRanges addObject:boldRanges.lastObject];
   }
-  index = 0;
+  
+  if (branch.remoteBranches.count > 0) {
+    [multilineTitle appendString:@"\n"];
+  }
+  
   for (GCHistoryTag* tag in branch.tags) {
-    if (index) {
-      [label appendString:@", "];
-    } else {
-      if (label.length) {
-        [label appendString:@"\n"];
-      }
-      [label appendString:@"["];
-    }
-    [label appendString:tag.name];
-    ++index;
+    NSString* tagName = tag.name;
+    [multilineTitle appendFormat:@"[%@]\n", tagName];
+    NSRange tagNameRange = NSMakeRange(multilineTitle.length - tagName.length - 2, tagName.length); // -2 to exclude char ']' plus '\n'
+    [boldRanges addObject:[NSValue valueWithRange:tagNameRange]];
+    [darkRanges addObject:boldRanges.lastObject];
   }
-  if (index) {
-    [label appendString:@"]"];
+  
+  if (branch.tags.count > 0) {
+    [multilineTitle appendString:@"\n"];
   }
-  if (!label.length) {
-    [label release];
-    return;  // This should only happen if we have a detached HEAD with no other references pointing to the commit
+  
+  if (multilineTitle.length == 0) {
+    [boldRanges release];
+    [darkRanges release];
+    [multilineTitle release];
+    return; // This should only happen if we have a detached HEAD with no other references pointing to the commit
   }
   
-  // Prepare text
+  // Cache attributes
+  static CTFontRef titleFont = NULL;
+  if (titleFont == NULL) {
+    titleFont = CTFontCreateUIFontForLanguage(kCTFontUIFontSystem, 12.0, CFSTR("en-US"));
+  }
+  static CTFontRef boldFont = NULL;
+  if (boldFont == NULL) {
+    boldFont = CTFontCreateUIFontForLanguage(kCTFontUIFontEmphasizedSystem, 12.0, CFSTR("en-US"));
+  }
+  static NSParagraphStyle* style = nil;
+  if (style == nil) {
+    NSMutableParagraphStyle* mutableStyle = [[NSMutableParagraphStyle alloc] init];
+    mutableStyle.lineHeightMultiple = 0.8;
+    style = [mutableStyle copy];
+    [mutableStyle release];
+  }
   
-  CFAttributedStringRef string = CFAttributedStringCreate(kCFAllocatorDefault, (CFStringRef)label, (CFDictionaryRef)attributes);
-  CTFramesetterRef framesetter = CTFramesetterCreateWithAttributedString(string);
-  CGSize size = CTFramesetterSuggestFrameSizeWithConstraints(framesetter, CFRangeMake(0, CFAttributedStringGetLength(string)), NULL, CGSizeMake(CGFLOAT_MAX, CGFLOAT_MAX), NULL);
+  // Create a light string to show above the HEAD
+  NSDictionary* multilineTitleAttributes = @{ NSFontAttributeName: (id)titleFont, NSForegroundColorAttributeName: [color shadowWithLevel:0.25], NSParagraphStyleAttributeName: style };
+  NSMutableAttributedString* multilineAttributedTitle = [[NSMutableAttributedString alloc] initWithString:multilineTitle attributes:multilineTitleAttributes];
+  
+  // Change font to bold on ranges collected before
```

Comment:
I didn't notice that before, but you don't need to create a `NSString` then keep around all the attributes, then create a `NSMutableAttributedString` recombining everything. Keep it simple: just do:
```
NSMutableAttributedString* multilineTitle = [[NSMutableAttributedString alloc] init];
(or if you need some base attributes for the entire string)
NSMutableAttributedString* multilineTitle = [[NSMutableAttributedString alloc] initWithString:@"" attributes:...];
[multilineTitle beginEditing];

(loop)
NSInteger length = multilineTitle.length;
[multilineTitle replaceCharactersInRange:NSMakeRange(length, 0) withString:...];
[multilineTitle setAttributes:... range:NSMakeRange(length, multilineTitle.length)];

[beginEditing endEditing];
// draw
[multilineTitle release];
```